Fix: use better form for a couple of translatable texts.

SlySven opened the pull request:
Note that this also adds to the small number of (plurals only) en_US translations - and several of the now 14 items there have been updated as well.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: use better form for a translatable text.

Note that this also adds to the small number of (plurals only) en_US
translations - and several of the now 14 items there have been updated as
well.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap reviewed (commented):
I don't favor the change to dlgPackageExporter.cpp - because you effectively changed a short text to a longer one:
"Select what to export"
"Select what to export (0 items already selected)"

The first one I like better, because it tells the user what to do.
The added info in the suggested version really adds no information the user didn't already know, just convolutes the message, making it easier to not read it at all.

The changes of mudlet_en_US.ts I don't understand. Where do they come from? Probably not in scope of this PR

vadi2 commented:
The text change just changes the code, it doesn't have an UI ramnifications - we already have this in 4.12. It's pretty handy as well, give it a try.

Kebap commented:
Wait but the code does change the UI text as well:

Before:
"Select what to export"

New:
"Select what to export (0 items already selected)"

vadi2 commented:
Both of the strings are already there - https://github.com/Mudlet/Mudlet/blob/Mudlet-4.12.0/src/dlgPackageExporter.cpp#L1433

![Peek 2021-08-10 16-46](https://user-images.githubusercontent.com/110988/128888209-7d6e61c0-0ef7-4d0b-b438-daa28f5af5a1.gif)

Very handy feature, for some large systems it helps you to know if you accidentally selected way too much in the export

vadi2 commented:
I see what you mean - indeed this change is for the worse, because this message is a bit silly:

![image](https://user-images.githubusercontent.com/110988/128896847-2c42d9d6-7c75-4119-a0c4-bd23d9ceb62b.png)

This happens after you select and then de-select something.

SlySven commented:
> This happens after you select and then de-select something.

That sounds like I did not spot that the original text is present in the `ui` file so is used initially. The rationale (as *I* saw it) was that we were special casing the "0 items" case (note that "item**s**" in English the zero quantity uses the same numerous plural form as anything other than unity - but other languages behave differently). I thought it made for simpler code to use the same engineering message for **all** quantities.

How about a compromise to use: `"What to export (%n item(s) selected):"` which removes the duplication of "select(ed)" in the text and, leading to, in English:

0. `"What to export (0 items selected):"`
1. `"What to export (1 item selected):"`
2. `"What to export (2 items selected):"`
3. `"What to export (3 items selected):"`
4. `"What to export (4 items selected):"`
5. `"What to export (5 items selected):"`

and I (think) in Russian:

0. `"Что экспортировать (0 выбранных предметов):"`
1. `"Что экспортировать (1 выбранный элемент):"`
2. `"Что экспортировать (2 выбранных элемента):"`
3. `"Что экспортировать (3 выбранных элемента):"`
4. `"Что экспортировать (4 выбранных элемента): "`
5. `"Что экспортировать (5 выбранных предметов):"`

> The changes of `mudlet_en_US.ts` I don't understand. Where do they come from? Probably not in scope of this PR

Very definitely within the scope - any time an engineering text in a `tr(...)` translatable string with a `%n` (filled in from the **third (`int`)** "count" argument) is add/modified the `mudlet_en_US.ts` file needs to be updated - but by virtue that it is a runt "plurals-only" translation file (after this PR it will have only 14 translations) it has to be done outside of the CrowdIn system which cannot special case one language like this.

vadi2 commented:
I don't think we should be saying there are 0 items selected when there's nothing selected, it's a bit too robotic. It's obvious to the person when there's nothing selected.

SlySven commented:
> I don't think we should be saying there are 0 items selected when there's nothing selected, it's a bit too robotic. It's obvious to the person when there's nothing selected.

:thinking: Okay - yeah there is something in what you say.

However we still should fix the non-zero case to use the correct (plural capable) translation - as it is, the existing output for a single item will be `"Select what to export (1 items)"` in English - which is wrong and probably more so for some other languages for other values...

SlySven committed:
Revise: changes prompted by peer-review

Revert to separate messages for some initial states when there is
nothing selected. This was requested by Kebap/Leris and was, I think,
what was stalling acceptance.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Agreed, let's do that.

vadi2 commented on src/dlgPackageExporter.cpp:
```diff
@@ -1427,11 +1427,11 @@ void dlgPackageExporter::slot_recountItems(QTreeWidgetItem *item)
-            } else {
-                mpSelectionText->setTitle(tr("Select what to export (%1 items)", "Package exporter selection", itemsToExport).arg(itemsToExport));
-            }
+            // Previously we special-cased the zero count case but in some
+            // locales (France, Poland, Russia, etc.) that is included in a
+            // plural form so it is better to just include it within the same
+            // translation as the other quantities:
+            mpSelectionText->setTitle(tr("Select what to export (%n item(s) already selected)", "Package exporter selection", itemsToExport));
```
```suggestion
            if (itemsToExport == 0) {
                mpSelectionText->setTitle(tr("Select what to export"));
            } else {
                mpSelectionText->setTitle(tr("Select what to export (%n item(s) already selected)", "Package exporter selection", itemsToExport));
            }
```

vadi2 committed:
Update src/dlgPackageExporter.cpp

vadi2 reviewed (approved):
LGTM... over to @Kebap

SlySven commented:
Hang on! Something seems to have gotten messed up here in 55481598ede2c1d965ddc38dc913911caffc2e2d:
```diff
             int itemsToExport = countCheckedItems();
-            // Previously we special-cased the zero count case but in some
-            // locales (France, Poland, Russia, etc.) that is included in a
-            // plural form so it is better to just include it within the same
-            // translation as the other quantities:
-            mpSelectionText->setTitle(tr("Select what to export (%n item(s) already selected)", "Package exporter selection", itemsToExport));
+            if (itemsToExport == 0) {
+                mpSelectionText->setTitle(tr("Select what to export"));
+            } else {
+                mpSelectionText->setTitle(tr("Select what to export (%n item(s) already selected)", "Package exporter selection", itemsToExport));
+            }
             debounce = false;
```
You have kept the overlong message with both "Select" and "selected" in it - I thought we wanted:
```diff
             int itemsToExport = countCheckedItems();
+            if (itemsToExport == 0) {
+                mpSelectionText->setTitle(tr("Select what to export"));
+            } else {
+                mpSelectionText->setTitle(tr("Select what to export (%n item(s))", "Package exporter selection", itemsToExport));
+            }
             debounce = false;
```
and whatever gets used needs to be `lupdate`d (and translations provided) into the `./translations/translated/mudlet_en_US.ts` file....

I'll just do that now!

vadi2 commented:
Indeed, thanks for that catch.

SlySven committed:
Redo: get the numerus translations right for the package exporter & manager

Note that the "Remove package" text on the button in the package manager
has a similar issue in that it would update the text to say
"Remove package" or "Remove packages" in English (American) but if any
selection was cleared in use it would not update the text to say that NO
packages were currently selected for removal (though it would be disabled).

I have tweaked things so that the total number of packages to be removed is
now shown - even when it is zero. This will also work better for other
locales where zero might be handled differently to one (not though, in
France!)

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
Fair enough with the original text. We could have merged that now. 
Instead, look at the scope creep with the latest commit.. 👀 

I am not happy with their new text phrasings either..
Would you like to discuss here or revert and roll a new PR?

SlySven commented on src/dlgPackageExporter.cpp:
```diff
@@ -1430,7 +1430,7 @@ void dlgPackageExporter::slot_recountItems(QTreeWidgetItem *item)
                 mpSelectionText->setTitle(tr("Select what to export"));
             } else {
-                mpSelectionText->setTitle(tr("Select what to export (%1 items)", "Package exporter selection", itemsToExport).arg(itemsToExport));
+                mpSelectionText->setTitle(tr("Select what to export (%n item(s))", "Package exporter selection", itemsToExport));
```
Note that this change correctly uses the way to get the plurals handled with a `%n`. So it is wanted.

SlySven commented on src/dlgPackageManager.cpp:
```diff
@@ -241,8 +241,5 @@ void dlgPackageManager::slot_toggle_remove_button()
-        // let the translations decide whenever it should be 'Remove package', 'Remove packages', or whatever is language-appropriate
-        mRemoveButton->setText(tr("Remove packages", "Button in package manager to remove selected package(s)", selectionCount));
-    }
+    mRemoveButton->setText(tr("Remove %n package(s)", "Button in package manager to remove selected package(s)", selectionCount));
```
This does what the original comment intended. By only updating the text when the selection was greater than one however the original code would not update the text should the number of packages to remove drop back to the (most likely usage) of 1 package - or if everything was deselected. So it is wanted.

Kebap replied:
Need to test in Mudlet, how this behaves when selectionCount = 0

SlySven replied:
it will show `Remove 0 packages` (at least in `en` locales) but will be in a light grey colour as it will also be disabled by the line before it.

Kebap replied:
> I don't think we should be saying there are 0 items selected when there's nothing selected, it's a bit too robotic. It's obvious to the person when there's nothing selected.

That applies here as well.

SlySven replied:
Doing that does make the 0 packages are selected a "special case" - and means we have to special case the text (and the translation) in the C++ code. I chose the other way, which tries to use the same text for 0 as for the other cases.

Kebap replied:
Yes, indeed, and that's the same discussion you had with @vadi2 over 5 weeks ago here above. Please make a special text for 0 items selected here as well.

SlySven commented on src/ui/package_manager.ui:
```diff
@@ -216,7 +216,7 @@
            </property>
            <property name="text">
-            <string>Remove packages</string>
+            <string comment="Initially nothing will be selected for removal so the count will be zero in the initial display of this text.">Remove 0 packages</string>
```
This ensures the initial text displayed in the button is the same as it is after something has been selected. So it is wanted.

Kebap replied:
Need to check in Mudlet. Seems weird to have a button say "Remove 0 packages". What happens upon clicking it? Nothing? Weird. Should probably not be clickable until packages get selected. Then the text doesn't need the 0.

SlySven replied:
It won't be enabled at that point i.e. initially, as it is disabled in the constructor IIRC - so indeed it won't be clickable - however after one or more packages have been selected it is also possible to deselected them ALL and return to the initial 0 number state.  Using this string in this way and in the C++ code just means that what is shown initially is identical to what is shown if the user returns to the initial state *after* interacting with the dialogue. To my mind it is a better UI experience if the same state produces the same visual (UI) output whether the user has done anything or not.

SlySven commented on translations/translated/mudlet_en_US.ts:
```diff
@@ -4,140 +4,90 @@
         <source>You have a map loaded (%n room(s)), but Mudlet does not know where you are at the moment.</source>
         <translation>
-            <numerusform>You have a map loaded (one room), but Mudlet does not know where you are at the moment.</numerusform>
+            <numerusform>You have a map loaded (1 room), but Mudlet does not know where you are at the moment.</numerusform>
```
The previous text said `one` for one room but `2` for two and so on. So the use of numbers/string needs to be fixed (IMHO). So this is wanted.

SlySven commented on translations/translated/mudlet_en_US.ts:
```diff
@@ -4,140 +4,90 @@
+        <location filename="../../src/TRoomDB.cpp" line="724"/>
         <source>[ ALERT ] - %n area(s) detected as missing in map: adding it/them in.
- Look for further messsages related to the rooms that are supposed
+ Look for further messages related to the rooms that are supposed
```
These are (probably my) typos that slipped into the source text and were cleaned up a long time ago - but the `mudlet_en_US.ts` file didn't get picked up upon - but now they have. So these changes are wanted.

SlySven commented on translations/translated/mudlet_en_US.ts:
```diff
@@ -4,140 +4,90 @@
 </context>
 <context>
-    <name>T2DMap</name>
+    <name>MapInfoContributorManager</name>
```
These next few changes are left over from the extension to the map info system - which was spun out to a new class. As such the `mudlet_en_US.ts` should have been updated then, but it wasn't until now. So this is wanted.

SlySven reviewed (commented):
>Fair enough with the original text. We could have merged that now.
>Instead, look at the scope creep with the latest commit.. :eyes:
>

In doing the original part I hit upon missing changes to a file that is outside the CrowdIn process. In updating that with the change made for the original bit it also pulled in changes that have accumulated (in the source "Engineering english" text) but which had not been done to that file.

>I am not happy with their new text phrasings either..

What parts are poorly phrased now?

>Would you like to discuss here or revert and roll a new PR?

Since I don't think there is anything *now* to revert I want to keep discussions here.

vadi2 commented:
@Kebap ?

Kebap commented:
> Since I don't think there is anything _now_ to revert I want to keep discussions here.

I meant you could revert this PR to its previous state, then merge that scope as it was green-lighted before.
The new changes could come in a new PR which would contain the additional discussion of the new items.
That way, the delay happening now could have been avoided and the existing changes merged instantly.

As it stands, I marked some changes with 👍 that seem uncritical to me, but didn't find time to comment in more detail on the others, yet. I am sorry this also blocks the other changes and that is why I explained the above.

Kebap commented on translations/translated/mudlet_en_US.ts:
```diff
@@ -4,140 +4,90 @@
+        <source>[ INFO ]  - The missing area(s) are now called:
+(ID) ==&gt; &quot;name&quot;</source>
+        <comment>Making use of %n to allow quantity dependent message form 8-) !</comment>
+        <translatorcomment>Perhaps should put the %n back in the source code</translatorcomment>
```
What does this mean then?

Kebap replied:
I see this was already commented before this commit so fair enough

Kebap commented:
Outlined the remaining checks above.

vadi2 commented:
This PR needs to be updated with latest `development` code to pull in the new mandatory Lua tests - would you be able to do that?

SlySven committed:
Merge branch 'development' into Revise_useBetterFormForATranslatableText

Update as this PR has been stalled for a while.

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
@vadi2 I believe this one is done (but a Bot thinks it is still *dangerous* because I didn't rename it until *after* that last commit was applied) OTOH with everything that is going on in the **4.15.0** release process I didn't want it to get into the PTB process by merging it into `development` yet. OTOH it does update the (manual) `en-US` translation so it might be beneficial to have it in that release after all!

vadi2 commented:
I'm good with it, @Kebap are you happy for this to be merged as well?

Kebap commented:
OK go ahead 👍
